src/python/arcor2_arserver/lock.py

```diff
@@ -0,0 +1,306 @@
+
+        if isinstance(obj_ids, str):
+            obj_ids = [obj_ids]
+
+        roots = [self.get_root_id(obj_id) for obj_id in obj_ids]
+        locked = []
+
+        async with self._lock:
+            for i, obj_id in enumerate(obj_ids):
+                lock_record = self._get_lock_record(roots[i])
+                acquired = lock_record.write_lock(obj_id, owner, lock_tree)
+
+                if acquired:
+                    locked.append(obj_id)
+                else:
+                    self._write_unlock(roots, locked, owner)
+                    return False
+        return True
+
+    async def write_unlock(self, obj_ids: Union[List[str], str], owner: str) -> None:
+
+        if isinstance(obj_ids, str):
+            obj_ids = [obj_ids]
+
+        roots = [self.get_root_id(obj_id) for obj_id in obj_ids]
+
+        async with self._lock:
+            self._write_unlock(roots, obj_ids, owner)
+
+    def _write_unlock(self, roots: List[str], obj_ids: List[str], owner: str) -> None:
+        """Internal function when lock is acquired"""
+
+        assert self._lock.locked()
+
+        for i, obj_id in enumerate(obj_ids):
+            try:
+                lock_record = self._get_lock_record(roots[i])
+                lock_record.write_unlock(obj_id, owner)
+
+                if lock_record.is_empty():
+                    del self._locked_objects[roots[i]]
+            except Exception:
+                self._write_unlock(roots[i+1:], obj_ids[i+1:], owner)
+                raise
+
+    async def is_write_locked(self, obj_id: str, owner: str) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._locked_objects.get(root_id)
+            if not lock_record:
+                return False
+
+            return obj_id in lock_record.write and owner in lock_record.write[obj_id].owners
+
+    async def is_read_locked(self, obj_id: str, owner: str) -> bool:
+
+        root_id = self.get_root_id(obj_id)
+
+        async with self._lock:
+            lock_record = self._locked_objects.get(root_id)
+            if not lock_record:
+                return False
+
+            return obj_id in lock_record.read and owner in lock_record.read[obj_id].owners
+
+    async def get_locked_roots_count(self) -> int:
+
+        async with self._lock:
+            return len(self._locked_objects)
+
+    def get_root_id(self, obj_id: str) -> str:
+        """Retrieve root object id for given object"""
+
+        if self.scene:
+            # TODO support scene object hierarchy
+            return obj_id
+
+        elif self.project:
+            obj = self.project.action_point(obj_id)
+            return obj.parent if obj.parent else obj_id
+
+        # TODO find out, how enum works
+        elif obj_id in [self.SpecialValues.SCENE_NAME, self.SpecialValues.PROJECT_NAME]:
+            return obj_id
+
+        else:
+            raise KeyError(f"Unknown object '{obj_id}'")
```

Comment:
Tady by bylo lepší vyvolávat něco odvozeného od Arcor2Exception. Máme to tak všude, že "naše" funkce a metody vyhazují jenom "naše" výjimky.